.github/workflows/ci-pipeline.yml

```
@@ -7,10 +7,57 @@ on:
   pull_request:
 
 jobs:
-  build:
+  Test_Build:
     name: Build and test
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
-    - run: docker build --target test --tag my-test-image .
-    - run: docker run --env-file .env.test my-test-image
+
+    - name: build test image
+      run: docker build --target test --tag my-test-image .
+
+    - name: Run tests
+      run: docker run --env-file .env.test my-test-image
+
+
+
+  Production_Build:
+    name: Build and Publish
+    if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
+    runs-on: ubuntu-latest    
+    needs: Test_Build
+
+    steps:
+      - uses: actions/checkout@v2
+        
+      - name: Build Production Image
+        run: docker build --target production --tag markrobertdunn/todo-app:$GITHUB_SHA .
+
+      - name: Login to Docker
+        run: docker login --username markrobertdunn --password ${{ secrets.DOCKERHUBKEY }}
+
+      - name: Publish Production Image to Docker
+        run: docker push markrobertdunn/todo-app:$GITHUB_SHA
+
+  Production_Release:
+    name: Release
+    runs-on: ubuntu-latest
+    needs: Production_Build
+    env: 
+      HEROKU_API_KEY: ${{secrets.HEROKU_API_KEY}}
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Build Production Image
+        run: docker build --target production --tag markrobertdunn/todo-app:$GITHUB_SHA .
```

Given that you build this image here solely to push to Heroku, we could skip the later tagging step and just tag it appropriately here straight off:
```suggestion
        run: docker build --target production --tag registry.heroku.com/devops-todoapp/web .
```